tests/requests/tests.py

```diff
@@ -752,6 +765,79 @@ def test_get_port_with_x_forwarded_port(self):
         }
         self.assertEqual(request.get_port(), '8080')
 
+    @override_settings(USE_X_FORWARDED_HOST=True)
+    def test_get_port_with_x_forwarded_host(self):
+        request = HttpRequest()
+        request.META = {
+            'SERVER_PORT': '8080',
+            'HTTP_X_FORWARDED_HOST': 'example.com:8000',
+        }
+        # Should use the X-Forwarded-Host header
+        self.assertEqual(request.get_port(), '8000')
+
+        request = HttpRequest()
+        request.META = {
+            'SERVER_PORT': '8080',
+        }
+        self.assertEqual(request.get_port(), '8080')
+
+    @override_settings(USE_X_FORWARDED_HOST=True)
+    def test_get_port_with_bad_x_forwarded_host(self):
+        request = HttpRequest()
+        request.META = {
+            'SERVER_PORT': '8080',
+            'HTTP_X_FORWARDED_HOST': 'example.com:',
+        }
+        # Should use the X-Forwarded-Host header
+        self.assertEqual(request.get_port(), '8080')
+
+    @override_settings(USE_X_FORWARDED_HOST=True,
+                       USE_X_FORWARDED_PORT=True)
+    def test_get_port_with_x_forwarded_host_and_port(self):
+        request = HttpRequest()
+        request.META = {
+            'SERVER_PORT': '8080',
+            'HTTP_X_FORWARDED_HOST': 'example.com',
+            'HTTP_X_FORWARDED_PORT': '8010',
+        }
+        # Should use the X-Forwarded-Port header
+        self.assertEqual(request.get_port(), '8010')
+
+        request = HttpRequest()
+        request.META = {
+            'SERVER_PORT': '8080',
+        }
+        self.assertEqual(request.get_port(), '8080')
+
+    @override_settings(
+        USE_X_FORWARDED_PORT=True,
+        USE_X_FORWARDED_HOST=True,
+        ALLOWED_HOSTS=['*'])
+    def test_get_host_with_x_forwarded_port(self):
+        request = HttpRequest()
+        request.META = {
+            'SERVER_PORT': '80',
+            'HTTP_X_FORWARDED_HOST': 'example.com',
+            'HTTP_X_FORWARDED_PORT': '8080',
+        }
+        # Should use the X-Forwarded-Port header
+        self.assertEqual(request.get_host(), 'example.com:8080')
+
+    @override_settings(
+        USE_X_FORWARDED_PORT=True,
+        USE_X_FORWARDED_HOST=True,
+        ALLOWED_HOSTS=['*'])
+    def test_get_host_with_use_x_forwarded_port_and_port_in_host(self):
+        request = HttpRequest()
+        request.META = {
+            'SERVER_PORT': '80',
+            'HTTP_X_FORWARDED_HOST': 'example.com:8081',
+            'HTTP_X_FORWARDED_PORT': '8080',
+        }
+        # Should use the X-Forwarded-Port header
+        with self.assertRaises(ImproperlyConfigured):
+            request.get_host()
+
```

Comment:
Excuse me my ignorance because I haven't been able to figure out exactly what the desired behaviour is when USE_X_FORWARDED_PORT=True but USE_X_FORWARDED_HOST=False, but specifically when using HTTP_X_FORWARDED_PORT in combination with HTTP_HOST, the current implementation seems to give inconsistent results.

Below I propose two additional tests which -with the current implementation- both pass, though maybe the second one shouldn't?

```suggestion

    @override_settings(
        USE_X_FORWARDED_PORT=True,
        USE_X_FORWARDED_HOST=False,
        ALLOWED_HOSTS=['*'])
    def test_get_host_with_use_x_forwarded_port_and_server_name(self):
        request = HttpRequest()
        request.META = {
            'SERVER_NAME': 'original.com',
            'SERVER_PORT': '80',
            'HTTP_X_FORWARDED_HOST': 'example.com',
            'HTTP_X_FORWARDED_PORT': '8080',
        }

        # Should use the X-Forwarded-Port header
        self.assertEqual(request.get_host(), 'original.com:8080')
        self.assertEqual(request.get_port(), '8080')

    @override_settings(
        USE_X_FORWARDED_PORT=True,
        USE_X_FORWARDED_HOST=False,
        ALLOWED_HOSTS=['*'])
    def test_get_host_with_use_x_forwarded_port_and_http_host(self):
        request = HttpRequest()
        request.META = {
            'HTTP_HOST': 'original.com',
            'SERVER_PORT': '8000',
            'HTTP_X_FORWARDED_HOST': 'example.com',
            'HTTP_X_FORWARDED_PORT': '8080',
        }

        # Should NOT use the X-Forwarded-Port header
        self.assertEqual(request.get_host(), 'original.com')
        self.assertEqual(request.get_port(), '8080')

```

Reply:
what is the difference between the first and second test?
by my interpretation there is nothing wrong with using `USE_X_FORWARDED_PORT` when not using `USE_X_FORWARDED_HOST` and since the headers may be outside of the control of the django user we should not throw errors when headers are present that we are not using (if that's the case you're indicating). Given that the headers should be independent when they don't conflict I think your test cases are equivalent to the `test_get_port_with_x_forwarded_port()` test case (tests/requests/tests.py +753).

Reply:
The fact that ```USE_X_FORWARDED_PORT``` can be used independently of ```USE_X_FORWARDED_HOST``` is actually the scenario I had in mind and it's exactly the convolutedness of the code, that I think @apollo13 is referring to, that makes it hard to reason about the various options and combinations (not intended as a critique on your contributions @dgcgh), so these tests are purely to demonstrate current behaviour.

To answer your question; yes my first suggested test does seem equivalent to ```test_get_port_with_x_forwarded_port()``` and is thus redundant. My second test considers the scenario where ```HTTP_HOST``` is specified in ```request.META``` instead of ```SERVER_NAME``` and I thought the two asserts might be contradictory (```get_host()``` returns ```'original'.com'``` thus implying port ```80``` while ```get_port()``` gives ```8080```).

Reply:
Yes, the second test should also yield `original.com:8080` from `get_host` because that is what the user originally opened in the browser (assuming the forwarded port got properly set).

Reply:
oic, yes, throughout this process I've been wishing the top level api followed the URL RFC for naming conventions: https://tools.ietf.org/html/rfc3986#section-3.2
`      authority   = [ userinfo "@" ] host [ ":" port ] `
there ought to be accessors for host and port which return just the host and port respectively, and perhaps a userinfo accessor, and then an authority accessor that is used to get the whole network layer address.
In my original use case, where we are attempting to reconstruct the use'rs [browser's] frame of reference to compose absolute URLs we really want to know the `authority`.

I think we're stuck with the semantics of `get_host()` returning both the `host` and `port`, but maybe we can make things a little cleaner.